### ISSUE TITLE
Invalid client_id

### ISSUE THREAD

**Issue by temp5374** [uptick/pymyob#45]
Hi everyone.

Firstly thanks for the library!

I am intermittently receiving an invalid client_id value error when processing the callback from MYOB.  It's a bit of a head scratcher as I don't have conditional logic so can't see why it's producing inconsistent results.  Often I will get the error on the first try, but then if I reload it will pass.  Then I wait a little while and try again and it fails, but on a reload it will pass.

I checked #21 but I can see that the PartnerCredentials.url is sending the client_id and I am on version  1.2.7 of pymyob.

Here's a simplified version of my processing view:

```
def get(self, request, *args, **kwargs):
    cred_state = cache.get('MYOB_creds')
    auth_secret = self.request.GET.get('code')
    creds = PartnerCredentials(**cred_state)
    creds.verify(auth_secret)
    myob = Myob(creds)
    files = myob.companyfiles.all()
    ....
```
```
File "/usr/local/lib/python3.6/dist-packages/oauthlib/oauth2/rfc6749/errors.py", line 405, in raise_from_error
    raise cls(**kwargs)
oauthlib.oauth2.rfc6749.errors.InvalidClientIdError: (invalid_request) Invalid client_id parameter value.
```

And inspecting the PartnerCredentials object
```
>>>creds.verify(auth_secret)
*** oauthlib.oauth2.rfc6749.errors.InvalidClientIdError: (invalid_request) Invalid client_id parameter value.
>>>creds.__dict__
{'consumer_key': '<>', 'consumer_secret': '<>', 'callback_uri': '<callback>', 'verified': True, 'companyfile_credentials': {}, 'oauth_token': '<token>', 'refresh_token': '<refresh>', 'oauth_expires_at': datetime.datetime(2020, 8, 17, 1, 36, 57, 324550), '_oauth': <requests_oauthlib.oauth2_session.OAuth2Session object at 0x7f5d50506780>, 'url': 'https://secure.myob.com/oauth2/account/authorize/?response_type=code&client_id=<client_id>&redirect_uri=<uri>&scope=CompanyFile'}
```

**Comment by jarekwg** [comment]
Hi @temp5374,

Thanks for the detailed report.
Just to confirm, this issue is solely around establishing the partner connection?
In the cases where you're saying it _does_ work, are you able then to proceed as normal from there? I'm curious whether other issues manifest afterwards.

Have not heard of or come across this before, so unfortunately don't have an answer for you presently; however would be interested in what you find if you keep digging!

**Comment by temp5374** [comment]
thanks for your reply :-)

The issue is solely around the initial .verify step in the flow.  On the times that the verification succeeds, I can save down the state and continue with `creds = PartnerCredentials(**cred_state)` and `Myob(creds)` and everything works great.

I initially thought I must be doing something silly with caching the state and it getting out of step (given it would not work the first time, but then works the second) but I seem to be having the same intermittent issue with that simplified case above so I'm not sure that's the issue.

Any thoughts on what I can inspect or what information I can provide to investigate further?

**Comment by jarekwg** [comment]
Hmm, "out of step" is an apt descrption of what could likely be happening. I probs won't be suggesting anything you haven't tried already, but:
- focus around the cached state -- clearly print out the whole dict before and after the myob round trip to ensure it's not gotten lost/changed in the interim.
- ensure you're not somehow calling things multiple times.
- ensure you're not waiting too long during the round trip.

Hopefully for your sake it's something simple like this; and not some obscure oauthlib issue deep down.
I can tell you we're using the latest `requests` and `requests-oauthlib` versions in production ourselves, so you can probs rule out it being something to do with more recent versions of these dependencies.

**Comment by jarekwg** [comment]
Closing. If you have more info, lemme know and will reopen.

**Comment by temp5374** [comment]
**TLDR; callback view is being called twice and the second time it fails.  Invalid client_id error appears to be a misnomer.**

Apologies for my radio silence (here and on the pull request!), I got pulled away into Xero land.  Thanks for spending the time on your detailed explanations.

Okay I've been looking at this in more depth over the last day and I think it is possible that your 'calling things multiple times' point above could be the culprit.  I'm not 100% on whether my interpretation is correct but a successful attempt looks like this:

```
[15/Sep/2020 23:45:24] "GET /link/myob/ HTTP/1.1" 302 0
[15/Sep/2020 23:45:31] "GET /link/myob/o/callback/?code=<code>&state=<state> HTTP/1.1" 302 0
[15/Sep/2020 23:45:34] "GET /link/myob/auth-file/ HTTP/1.1" 200 4263
```

And an unsuccessful attempt looks like this

```
[15/Sep/2020 23:56:56] "GET /link/myob/ HTTP/1.1" 302 0

Internal Server Error: /link/myob/o/callback/
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/dist-packages/django/core/handlers/exception.py", line 34, in inner
    response = get_response(request)
  File "/usr/local/lib/python3.6/dist-packages/django/core/handlers/base.py", line 126, in _get_response
    response = self.process_exception_by_middleware(e, request)
  File "/usr/local/lib/python3.6/dist-packages/django/core/handlers/base.py", line 124, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/opt/services/djangoapp/src/mounted-vol/link/views.py", line 383, in MyobProcessCallbackView
    creds.verify(auth_secret)
  File "/usr/local/lib/python3.6/dist-packages/myob/credentials.py", line 74, in verify
    include_client_id=True,
  File "/usr/local/lib/python3.6/dist-packages/requests_oauthlib/oauth2_session.py", line 360, in fetch_token
    self._client.parse_request_body_response(r.text, scope=self.scope)
  File "/usr/local/lib/python3.6/dist-packages/oauthlib/oauth2/rfc6749/clients/base.py", line 421, in parse_request_body_response
    self.token = parse_token_response(body, scope=scope)
  File "/usr/local/lib/python3.6/dist-packages/oauthlib/oauth2/rfc6749/parameters.py", line 431, in parse_token_response
    validate_token_parameters(params)
  File "/usr/local/lib/python3.6/dist-packages/oauthlib/oauth2/rfc6749/parameters.py", line 438, in validate_token_parameters
    raise_from_error(params.get('error'), params)
  File "/usr/local/lib/python3.6/dist-packages/oauthlib/oauth2/rfc6749/errors.py", line 405, in raise_from_error
    raise cls(**kwargs)
oauthlib.oauth2.rfc6749.errors.InvalidClientIdError: (invalid_request) Invalid client_id parameter value.

[15/Sep/2020 23:57:00] "GET /link/myob/o/callback/?code=<code>&state=<state> HTTP/1.1" 500 104083
[15/Sep/2020 23:57:02] "GET /link/myob/o/callback/?code=<code>&state=<state> HTTP/1.1" 302 0
```

It looks like maybe that when it fails it's because the callback view is being hit twice (and thus verify is being called twice).  Although that time stamp is out of order given my explanation, on other occasions, it looks like the first callback passes and then the second one 500's.

I can verify that indeed my ngrok is receiving two incomings to the callback view when it fails.  Strange that it appears to quite consistently occur every second time that I redirect to `credentials.url`.  It would seem more plausible than MYOB redirecting to me twice that I am redirecting to `credentials.url` twice, however I have a fairly straight forward view as below.

```
def myobstartauthview(request):
    credentials = PartnerCredentials(
        consumer_key=settings.MYOB_KEY,
        consumer_secret=settings.MYOB_SECRET,
        callback_uri=settings.MYOB_CALLBACK_URI,
    )

    cache.set(
        'MYOB_creds',
        credentials.state,
        timeout=None
    )

    return HttpResponseRedirect(credentials.url)
```

Currently I am handling this by just catching the error and if it occurs I just start the process again and redirect to credentials.url, but that's obviously very hacky!

Any thoughts on what might be going on?

**Comment by jarekwg** [comment]
Hmm, a view being hit twice..
So either MYOB itself is calling your endpoint twice, or something in your own setup might be.

- When you redirect to `credentials.url`, are you still being taken to a MYOB login page, or is it an immediate round trip back to your site (because you're already logged in with MYOB)? Not sure if this is relevant, but might help you find a pattern.
- Any chance there's double clicking going on somewhere along the way, that could trigger a request to run twice?
- Are all the query params fully formed in both the successful and failing calls?

I'm not sure what's happening for you -- can confirm that we only ever get a single request come back from MYOB in our own setup. If putting safeguards into your `o/callback/` endpoint, include some logging too and hopefully, the logs will shed some clues/patterns..
